Llama tutorial for TRTLLM (#62)
triton-inference-server/tutorials: Popular_Models_Guide/Llama2/trtllm_guide.md

#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#  * Neither the name of NVIDIA CORPORATION nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

> Note: This tutorial is for TensorRT-LLM Backend which is currently under development.

Author: Yes, because I think there might still be changes that come in TRT-LLM repo that might change this right?

> Note: This tutorial is for TensorRT-LLM Backend which is currently under development.

Reviewer: That's correct. It's just that `under development` might imply that the backend is not yet ready, maybe we could consider rewording it. Just my personal opinion, perfectly fine with keeping it as it is!

## Pre-build instructions

For this tutorial, we are using the Llama2-7B HuggingFace model with pre-trained weights.
Clone the repo of the model with weights and tokens [here](https://huggingface.co/meta-llama/Llama-2-7b-hf/tree/main). You will need to get permissions for the Llama2 repository as well as get access to the huggingface cli. To get access to the huggingface cli, go here: [huggingface.co/settings/tokens](https://huggingface.co/settings/tokens).

> ## Installation

Reviewer: I think it would be great to mention that they will have to update the submodule before launching the container:
```
# Update the submodules
cd tensorrtllm_backend
git submodule update --init --recursive
git lfs install
git lfs pull
```

Launch Triton docker container with TensorRT-LLM backend
```docker run --rm -it --net host --shm-size=2g --ulimit memlock=-1 --ulimit stack=67108864 --gpus all -v /path/to/tensorrtllm_backend:/tensorrtllm_backend nvcr.io/nvidia/tritonserver:23.10-trtllm-py3 bash```

Alternatively, you can follow instructions [here](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/README.md) to build Tritonserver with Tensorrt-LLM Backend if you want to build a specialized container.

Don't forget to allow gpu usage when you launch the container.

## Create Engines for each model [skip this step if you already have an engine]
TensorRT-LLM requires each model to be compiled for the configuration you need before running.

> To do so, before you run your model for the first time on Tritonserver you will need to create a TensorRT-LLM engine for the model for the configuration you want.

Reviewer: ```suggestion
To do so, before you run your model for the first time on Triton Server you will need to create a TensorRT-LLM engine for the model for the configuration you want.
```

To do so, you will need to complete the following steps:

1. Install Tensorrt-LLM python package
   ```bash
    # TensorRT-LLM is required for generating engines.
    pip install git+https://github.com/NVIDIA/TensorRT-LLM.git
    mkdir /usr/local/lib/python3.10/dist-packages/tensorrt_llm/libs/
    cp /opt/tritonserver/backends/tensorrtllm/* /usr/local/lib/python3.10/dist-packages/tensorrt_llm/libs/
    ```

> 

Reviewer: Just a side note that we would probably need to update this section for 23.11 release. We're removing some TRT-LLM libraries that are not required for the runtime, but needed for engine generation. So we need to run the build script to build TRT-LLM instead of pip installing the repo. The installation will look like below
```
function install_trt_llm {
    ARCH="$(uname -i)"
    # Install CMake
    wget https://github.com/Kitware/CMake/releases/download/v3.27.6/cmake-3.27.6-linux-${ARCH}.sh
    bash cmake-3.27.6-linux-*.sh --prefix=/usr/local --exclude-subdir && rm cmake-3.27.6-linux-*.sh
    export PATH="/usr/local/bin:${PATH}"

    # PyTorch needs to be built from soure for aarch64
    if [ "${ARCH}" = "aarch64" ]; then TORCH_INSTALL_TYPE="src_non_cxx11_abi"; \
    else TORCH_INSTALL_TYPE="pypi"; fi && \
    (cd /opt/tritonserver/tensorrtllm_backend/tensorrt_llm &&
        bash docker/common/install_pytorch.sh $TORCH_INSTALL_TYPE &&
        python3 ./scripts/build_wheel.py --trt_root="${TRT_ROOT}" &&
        pip3 install ./build/tensorrt_llm*.whl)
}
```
I can update the doc later for 23.11, created a ticket in my backlog.

2.  Log in to huggingface-cli

    ```bash
    huggingface-cli login --token hf_*****
    ```

> 3.  Compile model engines
>     The script to build Llama models is located in [TensorRT-LLM repository](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples). We use the one located in the docker container as

Reviewer: ```suggestion
3.  Compile model engines
    
    The script to build Llama models is located in [TensorRT-LLM repository](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples). We use the one located in the docker container as
```
nit: need extra line for formatting

     `/tensorrtllm_backend/tensorrt_llm/examples/llama/build.py`.

>      This command compiles the model with in flight batching and 1 GPU. More details for the scripting please see the documentation for the Llama example [here](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/llama/README.md).

Reviewer: nit
```suggestion
     This command compiles the model with inflight batching and 1 GPU. More details for the scripting please see the documentation for the Llama example [here](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/llama/README.md).
```

    ```bash
    python build.py --model_dir /<path to your llama repo>/Llama-2-7b-chat-hf/ \
                    --dtype bfloat16 \
                    --use_gpt_attention_plugin bfloat16 \
                    --use_inflight_batching \
                    --paged_kv_cache \
                    --remove_input_padding \
                    --use_gemm_plugin bfloat16 \

>                     --output_dir /<path to your engine>/1-gpu/

Reviewer 2: Missing `\` at end of line

>                     --world-size 1

Reviewer 2: ```suggestion
                    --world_size 1
```

    ```

    > Optional: You can check test the output of the model with `run.py`
    > located in the same llama examples folder.
    >
    >   ```bash
    >    python3 /run.py --engine_dir=<path to your engine>/1-gpu/ --max_output_len 100 --tokenizer_dir <path to your llama repo>/Llama-2-7b-chat-hf --input_text "How do I count to ten in French?"
    >    ```

## Serving with Triton

> We're almost there! The last step is to create a Triton readable model. You can

Reviewer 3: (Optional) To me "We're almost there!" reads as if the tutorial up to this point has taken an unreasonable amount of time. It may dissuade someone who is reading and not following the tutorial from giving it a try. But I may also be in my own head about this...your call :)

Author: It was mostly me complaining of how long it is. But it's a lot simpler now so I will remove this line

> find a template of a model that uses in flight batching in [tensorrtllm_backend/all_models/inflight_batcher_llm](https://github.com/triton-inference-server/tensorrtllm_backend/tree/main/all_models/inflight_batcher_llm).

Reviewer: nit
```suggestion
find a template of a model that uses inflight batching in [tensorrtllm_backend/all_models/inflight_batcher_llm](https://github.com/triton-inference-server/tensorrtllm_backend/tree/main/all_models/inflight_batcher_llm).
```

To run our Llama2-7B model, you will need to:

1. Copy over the inflight batcher models repository
 ```bash
 cp -R /tensorrtllm_backend/all_models/inflight_batcher_llm /opt/tritonserver/.
 ```

> 2. Modify config.pbtxt for the preprocessing, postprocessing and processing steps

Reviewer: I think maybe we could point to the documentation for more details for these models and tags: https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/README.md#create-the-model-repository

    ```bash
    # preprocessing
    sed -i 's#${tokenizer_dir}#/<path to your engine>/1-gpu/#' /opt/tritonserver/inflight_batcher_llm/preprocessing/config.pbtxt
    sed -i 's#${tokenizer_type}#auto#' /opt/tritonserver/inflight_batcher_llm/preprocessing/config.pbtxt
    sed -i 's#${tokenizer_dir}#/<path to your engine>/1-gpu/#' /opt/tritonserver/inflight_batcher_llm/postprocessing/config.pbtxt
    sed -i 's#${tokenizer_type}#auto#' /opt/tritonserver/inflight_batcher_llm/postprocessing/config.pbtxt

    sed -i 's#${decoupled_mode}#false#' /opt/tritonserver/inflight_batcher_llm/tensorrt_llm/config.pbtxt
    sed -i 's#${engine_dir}#/<path to your engine>/1-gpu/#' /opt/tritonserver/inflight_batcher_llm/tensorrt_llm/config.pbtxt
    ```
    Also, ensure that the `gpt_model_type` parameter is set to `inflight_fused_batching`

3.  Launch Tritonserver

    ```bash
    tritonserver --model-repository=/opt/tritonserver/inflight_batcher_llm
    ```

## Client

> You can test the results of the run with the [inflight_batcher_llm_client.py script](https://github.com/triton-inference-server/tensorrtllm_backend/tree/main/inflight_batcher_llm)

Reviewer: Maybe we could also demonstrate using the [generate endpoint](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/README.md#query-the-server-with-the-triton-generate-endpoint)? Open to discussion as we would need to mention that the endpoint is only in 23.10 and later releases. If users are using the Triton TRT-LLM container then it's fine, but if they are building from [dockerfile](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/README.md#option-3-build-via-docker) then it's still using Triton 23.08.

    The script to build Llama models is located in [TensorRT-LLM repository](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples). We use the one located in the docker container as `/tensorrtllm_backend/tensorrt_llm/examples/llama/build.py`.
    This command compiles the model with inflight batching and 1 GPU. To run with more GPUs, you will need to change the build command to use `--world_size X`.
    More details for the scripting please see the documentation for the Llama example [here](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/llama/README.md).

    ```bash
    python build.py --model_dir /<path to your llama repo>/Llama-2-7b-hf/ \
                    --dtype bfloat16 \
                    --use_gpt_attention_plugin bfloat16 \
                    --use_inflight_batching \
                    --paged_kv_cache \
                    --remove_input_padding \
                    --use_gemm_plugin bfloat16 \
                    --output_dir /<path to your engine>/1-gpu/ \
                    --world_size 1
    ```

    > Optional: You can check test the output of the model with `run.py`
    > located in the same llama examples folder.
    >
    >   ```bash

>     >    python3 /run.py --engine_dir=<path to your engine>/1-gpu/ --max_output_len 100 --tokenizer_dir <path to your llama repo>/Llama-2-7b-hf --input_text "How do I count to ten in French?"

Reviewer 3: ```suggestion
    >    python3 run.py --engine_dir=<path to your engine>/1-gpu/ --max_output_len 100 --tokenizer_dir <path to your llama repo>/Llama-2-7b-hf --input_text "How do I count to ten in French?"
```

 ```

2. Modify config.pbtxt for the preprocessing, postprocessing and processing steps. See details in [documentation](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/README.md#create-the-model-repository):

    ```bash
    # preprocessing
    sed -i 's#${tokenizer_dir}#/<path to your engine>/1-gpu/#' /opt/tritonserver/inflight_batcher_llm/preprocessing/config.pbtxt
    sed -i 's#${tokenizer_type}#auto#' /opt/tritonserver/inflight_batcher_llm/preprocessing/config.pbtxt
    sed -i 's#${tokenizer_dir}#/<path to your engine>/1-gpu/#' /opt/tritonserver/inflight_batcher_llm/postprocessing/config.pbtxt
    sed -i 's#${tokenizer_type}#auto#' /opt/tritonserver/inflight_batcher_llm/postprocessing/config.pbtxt

    sed -i 's#${decoupled_mode}#false#' /opt/tritonserver/inflight_batcher_llm/tensorrt_llm/config.pbtxt
    sed -i 's#${engine_dir}#/<path to your engine>/1-gpu/#' /opt/tritonserver/inflight_batcher_llm/tensorrt_llm/config.pbtxt
    ```
    Also, ensure that the `gpt_model_type` parameter is set to `inflight_fused_batching`

3.  Launch Tritonserver

    ```bash
    tritonserver --model-repository=/opt/tritonserver/inflight_batcher_llm
    ```

>     Note if you built the engine with `--world-size X` where `X` is greater than 1, you will need to use the [launch_triton_server.py](https://github.com/triton-inference-server/tensorrtllm_backend/blob/release/0.5.0/scripts/launch_triton_server.py) script.

Reviewer 3: ```suggestion
    Note if you built the engine with `--world_size X` where `X` is greater than 1, you will need to use the [launch_triton_server.py](https://github.com/triton-inference-server/tensorrtllm_backend/blob/release/0.5.0/scripts/launch_triton_server.py) script.
```
